### PR TITLE
Implemented Costume Biz Scenario

### PR DESCRIPTION
**Pull request by ChavaRaitzik**
I implemented the Costume Biz Scenario.
Please review and advise if there are any corrections needed.

### CODE REVIEW

File: 1 Table Customer.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
+use CostumeBusinessDB
+go 
+drop table if exists dbo.Customer
+go 
+create table dbo.Customer(
+    CustomerID int not null identity primary key,
```
Comment (AvigailGellis) [inline 1 Table Customer.sql:6]: Syntax tip: column names keep camel case ie: CustomerId

File: 1 Table Customer.sql [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        check(CostumeBought in ('American Girl Doll', 'Artist', 'Bumble Bee', 'Colonial Boy', 'Colonial Girl', 'Elephant', 'Fire Man', 'Police Man', 'Princess', 'Zebra')),
+    CostumeSize varchar(2) not null 
+        constraint ck_Customer_Costume_Size_must_be_XS_S_M_L_or_XL check(CostumeSize in ('XS','S','M','L','XL')),
+    AmountBought int not null 
+        constraint ck_Customer_Amount_Bought_must_be_greater_than_0 check(AmountBought > 0), 
+    CostPricePerCostume decimal(4,2) not null 
+        constraint ck_Customer_Cost_Price_Per_Costume_must_be_15_17_20_22_25 check(CostPricePerCostume in (15,17,20,22,25)),
+    SoldPricePerCostume decimal(4,2) not null,
+    FullPrice bit not null, 
+    DateBought date not null,
```
Comment (AvigailGellis) [inline 1 Table Customer.sql:22]: Add constraint that DateBought has to be after the business opened and not in the future.

File: 1 Table Customer.sql [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        constraint ck_Customer_Costume_Bought_must_be_a_costume_from_the_list_of_costumes_sold
+        check(CostumeBought in ('American Girl Doll', 'Artist', 'Bumble Bee', 'Colonial Boy', 'Colonial Girl', 'Elephant', 'Fire Man', 'Police Man', 'Princess', 'Zebra')),
+    CostumeSize varchar(2) not null 
+        constraint ck_Customer_Costume_Size_must_be_XS_S_M_L_or_XL check(CostumeSize in ('XS','S','M','L','XL')),
+    AmountBought int not null 
+        constraint ck_Customer_Amount_Bought_must_be_greater_than_0 check(AmountBought > 0), 
+    CostPricePerCostume decimal(4,2) not null 
+        constraint ck_Customer_Cost_Price_Per_Costume_must_be_15_17_20_22_25 check(CostPricePerCostume in (15,17,20,22,25)),
+    SoldPricePerCostume decimal(4,2) not null,
+    FullPrice bit not null,
```
Comment (AvigailGellis) [inline 1 Table Customer.sql:21]: For FullPrice add that if cost and sold price are the same then 0, if not then 1.

File: 1 Table Customer.sql [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    FullPrice bit not null, 
+    DateBought date not null,
+    DateSold date not null 
+        constraint ck_Customer_Date_Sold_must_be_before_or_on_the_current_date check(DateSold <= getdate()),
+    CustomerTotal as SoldPricePerCostume * AmountBought persisted,
+    TotalProfit as (SoldPricePerCostume - CostPricePerCostume) * AmountBought persisted,
+    constraint ck_Customer_Date_Bought_must_be_between_01_01_2020_and_Date_Sold check(DateBought between '01/01/2020' and DateSold),
+    constraint ck_Customer_Sold_Price_Per_Costume_must_be_between_Cost_Price_Per_Costume_and_30 check(SoldPricePerCostume between CostPricePerCostume and 30), 
+    constraint ck_Customer_FullPrice_is_1_only_when_SoldPricePerCostume_does_not_equal_CostPricePerCostume
+    check((SoldPricePerCostume = CostPricePerCostume and FullPrice = 0) or (SoldPricePerCostume <> CostPricePerCostume and FullPrice = 1))
```
[the remark this answers is not in the document]
Comment (AvigailGellis) [reply 1 Table Customer.sql:30]: Add a constraint that checks that costume sizes and cost per costume matches up according to the owner's instructions.

File: 2 Data Customer.sql [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	Yisroel Horowitz, Police Man, XL, 1, 30, Jan 16, 2022, Jan 19, 2022
+	Aliza Duetch, American Girl Doll, S, 2, 22, Mar 12, 2021, June 21, 2021
+	Rochel Rubin, Bumble Bee, S, 1, 22, Sep 11, 2020, Jan 2, 2021
+	Bracha Ganz, Princess, M, 4, 25, Nov 3, 2020, Dec 12, 2021
+	Yaakov Cohen, Princess, XS, 1, 20, Dec 4, 2021, July 25, 2022
+	Rina Rosen, Artist, M, 1, 25, Feb 18, 2022, May 28, 2022
+	Rivkah Goldberger, zebra, S, 1, 22, Sep 14, 2022, Dec 29, 2022 
+	*/
+
+Insert Customer(CustomerFirstName, CustomerLastName, CostumeBought, CostumeSize, AmountBought, CostPricePerCostume, SoldPricePerCostume, FullPrice, DateBought, DateSold)
```
Comment (AvigailGellis) [inline 2 Data Customer.sql:31]: ake out FullPrice once you change it to a computed column

File: 1 Table Customer.sql [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        constraint ck_Customer_Amount_Bought_must_be_greater_than_0 check(AmountBought > 0), 
+    CostPricePerCostume decimal(4,2) not null 
+        constraint ck_Customer_Cost_Price_Per_Costume_must_be_15_17_20_22_25 check(CostPricePerCostume in (15,17,20,22,25)),
+    SoldPricePerCostume decimal(4,2) not null,
+    FullPrice bit not null, 
+    DateBought date not null,
+    DateSold date not null 
+        constraint ck_Customer_Date_Sold_must_be_before_or_on_the_current_date check(DateSold <= getdate()),
+    CustomerTotal as SoldPricePerCostume * AmountBought persisted,
+    TotalProfit as (SoldPricePerCostume - CostPricePerCostume) * AmountBought persisted,
```
Comment (AvigailGellis) [inline 1 Table Customer.sql:26]: The client didn't ask for this.

File: 1 Table Customer.sql [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    AmountBought int not null 
+        constraint ck_Customer_Amount_Bought_must_be_greater_than_0 check(AmountBought > 0), 
+    CostPricePerCostume decimal(4,2) not null 
+        constraint ck_Customer_Cost_Price_Per_Costume_must_be_15_17_20_22_25 check(CostPricePerCostume in (15,17,20,22,25)),
+    SoldPricePerCostume decimal(4,2) not null,
+    FullPrice bit not null, 
+    DateBought date not null,
+    DateSold date not null 
+        constraint ck_Customer_Date_Sold_must_be_before_or_on_the_current_date check(DateSold <= getdate()),
+    CustomerTotal as SoldPricePerCostume * AmountBought persisted,
```
Comment (AvigailGellis) [inline 1 Table Customer.sql:25]: For those that are paying cost price the calculation will be different.

File: 3 Reports Customer.sql [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+group by c.CostumeSize
+order by NumOfSizesBought desc
+
+--3)
+select CustomerSummary = concat(c.CustomerFirstName, ' ', c.CustomerLastName, ': ', c.AmountBought, ' - ', c.CostumeBought, ' (', c.CustomerTotal, ')')
+from Customer c
+order by c.CustomerFirstName
+
+--4)
+select c.CustomerFirstName, c.CustomerLastName, c.TotalProfit
```
Comment (AvigailGellis) [inline 3 Reports Customer.sql:29]: Total profit is being taken out so change this to the correct calcuation for total profit

File: 1 Table Customer.sql (hunk 1 above)
Comment (ChavaRaitzik) [reply 1 Table Customer.sql:6]: Thanks, I corrected this to CustomerId.

File: 1 Table Customer.sql (hunk 3 above)
Comment (ChavaRaitzik) [reply 1 Table Customer.sql:21]: Thanks, I corrected this to a computed column.

File: 1 Table Customer.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChavaRaitzik) [reply 1 Table Customer.sql:30]: Thanks, I made FullPrice a computed column and took out the constraint.

I added the constraint for the costume sizes and costs.

File: 2 Data Customer.sql (hunk 5 above)
Comment (ChavaRaitzik) [reply 2 Data Customer.sql:31]: Removed FullPrice.

File: 1 Table Customer.sql (hunk 6 above)
Comment (ChavaRaitzik) [reply 1 Table Customer.sql:26]: Removed the column.

File: 1 Table Customer.sql (hunk 7 above)
Comment (ChavaRaitzik) [reply 1 Table Customer.sql:25]: Sorry, I don't understand why the calculation would be different. 

I did it according to the SoldPricePerCostume, so it should work. (Customers who pay Cost Price have the Cost Price in as their Sold Price.)

File: 3 Reports Customer.sql (hunk 8 above)
Comment (ChavaRaitzik) [reply 3 Reports Customer.sql:29]: Thanks, added the correct calculation

**Comment by ChavaRaitzik** [comment]
Please disregard this comment. I uploaded the wrong files.

**Comment by ChavaRaitzik** [comment]
Thank you. 
I made the corrections. Please review and advise if it's good now.

File: 1 Table Customer.sql (hunk 7 above)
Comment (AvigailGellis) [reply 1 Table Customer.sql]: Yes, that was my mistake

**Pull request closed without merging**